### PR TITLE
Web (HTTP) Seeding

### PR DESCRIPTION
**Pull request by olalonde**
WIP (do not merge). See https://github.com/feross/webtorrent/issues/67
Depends on this PR: https://github.com/feross/bittorrent-swarm/pull/19

I started working on a web seed implementation. The implementation right now is a bit naive (it maps every bittorrent protocol requests to HTTP `Range` requests) and could probably be optimized.

- [x] downloads from web seeds listed in `url-list`
- [ ] test in web browser
- [ ] test what happens when regular peers and web peer

### CODE REVIEW

**Comment by olalonde** [comment]
@feross could you just take a quick look at this and let me know if there's any chance to get this merged? And if I'm heading in the right direction...

**Comment by fbodz** [comment]
@feross this would be quite handy.

File: lib/torrent.js [hunk 1]
```diff
@@ -403,6 +407,16 @@ Torrent.prototype.addPeer = function (peer) {
 }
 
 /**
+ * Add a web peer to the swarm
+ * @param {string} web peer url
```
Comment (feross) [inline lib/torrent.js]: The correct syntax is:

```
@param {string} variableName description
```

**Comment by feross** [comment]
@olalonde - you're a hero! Thanks for this PR! Please accept my sincere apologies for the delay in getting this merged. This was a much needed feature. 

It only works in node right now, and it only supports BEP19, and the approach isn't very efficient with so many separate requests, but that's all okay for now. We have a starting point to work from now!

**Comment by feross** [comment]
This is released as 0.51.0.

**Comment by olalonde** [comment]
Great, @feross. Was starting to wonder if you'd ever take a look at the PR :+1:. I'll probably try to improve it in coming days/weeks as I need the functionality for a project I'm working on. I'll try to catch you on IRC if I need some advice.

**Comment by feross** [comment]
Sounds great!

**Comment by samcruz** [comment]
Hi there! I'm also interested in support for web seeding. I didn't try this out yet, but from reading the code, it seems this only supports single-file torrents, is that correct? Because if I'm reading BEP19 correctly, one would need to translate from pieces to offsets into specific files for multi-file torrents, which I don't see being done here.

**Comment by feross** [comment]
@samcruz Yeah, that's not implemented yet.

**Comment by batjko** [comment]
Does this mean, the testing indicated above hasn't been done before merging?

**Comment by jakefb** [comment]
Great work @olalonde! I'm trying to test this out but I'm getting an error

When I run the test file `download-webseed-torrent.js` in node I get:

    TAP version 13
    # Download using webseed (via .torrent file)

    TypeError: Cannot read property 'length' of undefined
        at WebConn.Wire.handshake (/home/jake/webtorrent/node_modules/bittorrent-swarm/node_modules/bittorrent-protocol/index.js:190:39)
        at WebConn.<anonymous> (/home/jake/webtorrent/node_modules/bittorrent-swarm/lib/webconn.js:26:10)
        at WebConn.emit (events.js:106:17)
        at WebConn.Wire._onHandshake (/home/jake/webtorrent/node_modules/bittorrent-swarm/node_modules/bittorrent-protocol/index.js:391:8)
        at WebConn.<anonymous> (/home/jake/webtorrent/node_modules/bittorrent-swarm/node_modules/bittorrent-protocol/index.js:665:12)
        at WebConn.Wire._write (/home/jake/webtorrent/node_modules/bittorrent-swarm/node_modules/bittorrent-protocol/index.js:555:10)
        at doWrite (_stream_writable.js:226:10)
        at writeOrBuffer (_stream_writable.js:216:5)
        at WebConn.Writable.write (_stream_writable.js:183:11)
        at write (_stream_readable.js:585:24)

**Comment by feross** [comment]
@jakefb Hey! Were you by chance using a magnet link when you got that error? I fixed an issue with that in https://github.com/feross/webtorrent/commit/cb5c9a430c53d5cfd5e7db2ef2e0bdba7ec2013d and released it as 0.53.4. Please update and give it another try!

**Pull request merged**